src/client/action/list_databases.rs

```diff
@@ -0,0 +1,140 @@
+use std::{future::IntoFuture, marker::PhantomData};
+
+use bson::{Document, Bson};
+use futures_util::FutureExt;
+
+use crate::{Client, ClientSession, operation::list_databases as op, error::{Result, ErrorKind}, results::DatabaseSpecification, client::BoxFuture};
+
+impl Client {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(
+        &self,
+    ) -> ListDatabases {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+
+    /// Gets the names of the databases present in the cluster the Client is connected to.
+    pub fn list_database_names(
+        &self,
+    ) -> ListDatabases<'_, Names> {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+}
+
+/// Gets information about each database present in the cluster the Client is connected to.
+#[must_use]
+pub struct ListDatabases<'a, M: Mode = Full> {
+    client: &'a Client,
+    options: Option<op::Options>,
+    session: Option<&'a mut ClientSession>,
+    mode: PhantomData<M>,
+}
+
+
+mod private {
```

Comment:
can you explain what the trait is needed for?

Reply:
The `Mode` trait has to be public because it's part of the API.  However, we don't want users adding impls of `Mode` to new types because it's just there to allow the two different paths for `ListDatabases`.  This hack is the typical way to do that in Rust: https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed